File: android/app/src/main/java/es/manabe/yomiyasu/core/models/DictionaryModels.kt
package es.manabe.yomiyasu.core.models

import kotlinx.serialization.Serializable

@Serializable
data class DictionaryDisplay(
    val display: String? = null,
    val words: List<DictionaryWord> = emptyList(),
)

@Serializable
data class DictionaryWord(
    val id: String,
    val frequency: String? = null,
    val pitches: List<DictionaryPitch>? = null,
    val kanji: List<DictionaryKanji>? = null,
    val kana: List<DictionaryKana>? = null,
    val sense: List<DictionarySense>? = null,
) {
    val headword: String
        get() = kanji?.firstOrNull()?.text ?: kana?.firstOrNull()?.text ?: ""

    val mainReading: String
        get() = kana?.firstOrNull()?.text ?: ""

    val pitchPositions: List<Int>
        get() = pitches?.map { it.position } ?: emptyList()

    val frequencyRank: Int?
        get() = frequency?.toIntOrNull()

    val firstGlosses: List<String>
        get() = sense?.firstOrNull()?.gloss?.map { it.text } ?: emptyList()
}

@Serializable
data class DictionaryPitch(val position: Int)

@Serializable
data class DictionaryKanji(
    val common: Boolean? = null,
    val text: String,
    val tags: List<String>? = null,
)

@Serializable
data class DictionaryKana(
    val common: Boolean? = null,
    val text: String,
    val tags: List<String>? = null,
    val appliesToKanji: List<String>? = null,
)

@Serializable
data class DictionarySense(
    val partOfSpeech: List<String>? = null,
    val appliesToKanji: List<String>? = null,
    val appliesToKana: List<String>? = null,
    val misc: List<String>? = null,
    val gloss: List<DictionaryGloss>? = null,
)

@Serializable
data class DictionaryGloss(
    val lang: String? = null,
    val text: String,
)

@Serializable
data class UserWordRequest(
    val word: String,
    val display: String,
    val sentence: String,
    val meaning: List<String>,
    val reading: String,
    val frequency: Double,
    val pitch: List<Int>,
)

@Serializable
data class SaveWordResponse(val modifiedCount: Int = 0)

File: ios/Yomiyasu/Core/Models/DictionaryModels.swift
import Foundation

struct DictionaryDisplay: Decodable, Sendable, Equatable {
    let display: String?
    let words: [DictionaryWord]
}

struct DictionaryWord: Decodable, Sendable, Equatable, Identifiable {
    let id: String
    let frequency: String?
    let pitches: [DictionaryPitch]?
    let kanji: [DictionaryKanji]?
    let kana: [DictionaryKana]?
    let sense: [DictionarySense]?
}

struct DictionaryPitch: Decodable, Sendable, Equatable {
    let position: Int
}

struct DictionaryKanji: Decodable, Sendable, Equatable {
    let common: Bool?
    let text: String
    let tags: [String]?
}

struct DictionaryKana: Decodable, Sendable, Equatable {
    let common: Bool?
    let text: String
    let tags: [String]?
    let appliesToKanji: [String]?
}

struct DictionarySense: Decodable, Sendable, Equatable {
    let partOfSpeech: [String]?
    let appliesToKanji: [String]?
    let appliesToKana: [String]?
    let misc: [String]?
    let gloss: [DictionaryGloss]?
}

struct DictionaryGloss: Decodable, Sendable, Equatable {
    let lang: String?
    let text: String
}

extension DictionaryWord {
    var headword: String {
        kanji?.first?.text ?? kana?.first?.text ?? ""
    }

    var mainReading: String {
        kana?.first?.text ?? ""
    }

    var pitchPositions: [Int] {
        pitches?.map(\.position) ?? []
    }

    var frequencyRank: Int? {
        frequency.flatMap(Int.init)
    }

    var firstGlosses: [String] {
        sense?.first?.gloss?.map(\.text) ?? []
    }
}

struct UserWordRequest: Encodable, Sendable {
    let word: String
    let display: String
    let sentence: String
    let meaning: [String]
    let reading: String
    let frequency: Double
    let pitch: [Int]
}
